### PR TITLE
Allow specifying parent project for autocreated projects (bom upload)

### LINKED ISSUE #2412: API /v1/bom should have a parameter parent
### Current Behavior

At the moment the API does not hava a parameter to specifiy the parent, when uploading a bom and using projectname/version.

### Proposed Behavior

Add a parameter "parent" to /v1/bom which consumes and sets the parant on upload.

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

### PR DESCRIPTION
**Pull request by Hunroll**
Added 3 params:
parentName, parentVersion, parentUUID.

If parentUUID is specified, it is used to find parent project and attach it to newly created one. If not, parentName and parentVersion are used. If parentUUID or parentName were specified but such project does not exist, 404 is returned.

### Description

Extended POST /api/v1/bom (uploadBom) with 3 optional params: parentName, parentVersion, parentUUID.
They are used when new project is created (autocreate=true) to attach it to specified project. If parent UUID and name are not specified, everything works as before.

### Addressed Issue

Closes #2412 (API /v1/bom should have a parameter parent)

### Additional Details

Our usecase:
We have some "root" projects (proj1, proj2, proj3) which we want to see on the dashboard and which have different policies.
And we have a huge number of versions (proj1:1.0.0, proj1:1.0.1, ....), all of them have their parent set to unversioned proj1.

So we had to extend API to be able to upload new sbom versions like that:
`curl -X "POST" $BASE_URL/api/v1/bom -H "Content-Type: multipart/form-data" -H "X-API-Key: $TOKEN" -F "autoCreate=true" -F "projectName=proj1" -F "projectVersion=2.0.1" -F "parentName=proj1" -F "bom=@bom.json"`
You can also explicitly specify version:
`-F "parentVersion="   or   -F "parentVersion=1.0.0"`
Or you can just use parent's UUID:
`curl -X "POST" $BASE_URL/api/v1/bom -H "Content-Type: multipart/form-data" -H "X-API-Key: $TOKEN" -F "autoCreate=true" -F "projectName=proj1" -F "projectVersion=2.0.1" -F "parenUUID=8524db04-7457-4c28-acdd-a4f5a4577d01" -F "bom=@bom.json"`

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [~] This PR fixes a defect, and I have provided tests to verify that the fix is effective
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [~] This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)
- [x] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly

### CODE REVIEW

File: src/main/java/org/dependencytrack/resources/v1/BomResource.java [hunk 1]
```diff
@@ -266,7 +269,21 @@ public Response uploadBom(@FormDataParam("project") String projectUuid,
...
+                              parent = qm.getObjectByUuid(Project.class, parentUUID);
+                            } else {
+                              final String trimmedParentName = StringUtils.trimToNull(parentName);
+                              final String trimmedParentVersion = StringUtils.trimToNull(parentVersion);
+                              parent = qm.getProject(trimmedParentName, trimmedParentVersion);
+                            }
+
+                            if (parent == null) { // if parent project is specified but not found
+                                return Response.status(Response.Status.NOT_FOUND).entity("The parent component could not be found.").build();
+                            }
```
[the remark this answers is not in the document]
Comment (Hunroll) [reply src/main/java/org/dependencytrack/resources/v1/BomResource.java:287]: Thanks for your comments!
Access control for parent project was added.
Tests are on the way

**Comment by Hunroll** [comment]
@nscuro Niklas, there are some problems with the tests. PUT method is covered, but POST is not. And currently I've extended only POST method.
What do you think if I merge PUT and POST methods, so they use the same code, just with different way to initialize a model. This way we can cover both workflows (PUT and POST) with same tests. 
As I see now they are almost identical, except other form of input validation.

Long story short, I want to call uploadBom PUT from uploadBom POST.
Thanks!

**Comment by valentijnscholten** [comment]
Nice, it makes scripting easier for CI/CD etc.

Just thinking out loud. I saw in some other PR that the `parentUUID` field has been removed from the Project json data as it is now modeled as `parent: { uuid: "..." }`. I swa thinking it should be done the same way here, but it looks like it's not possible. There's the form post endpoint which doesn't handle json payloads. And the endpoint that actually handles json already accepts `projectName`, `projectVersion` as flat json fields.

**Comment by Hunroll** [comment]
> Nice, it makes scripting easier for CI/CD etc.
> 
> Just thinking out loud. I saw in some other PR that the `parentUUID` field has been removed from the Project json data as it is now modeled as `parent: { uuid: "..." }`. I swa thinking it should be done the same way here, but it looks like it's not possible. There's the form post endpoint which doesn't handle json payloads. And the endpoint that actually handles json already accepts `projectName`, `projectVersion` as flat json fields.

Hello Valentijn!
Thanks for your comment!
Initially I've extended only the POST method, so it worked fine with additional form fields.
For PUT method I had to extend the org.dependencytrack.resources.v1.vo.BomSubmitRequest (not a project model, but a submit request), and it also works fine now. Also it allowed to cover this functionality with tests.

Thank you!

**Review by nscuro — APPROVED** [review]
